test/reduce-e2e/reduce_test.go

```diff
@@ -31,8 +31,118 @@ type ReduceSuite struct {
+	w.Expect().VertexPodLogContains("sink", "Payload -  20  Key -  even  Start -  60000  End -  70000")
+}
+
+// one reduce vertex(non keyed)
+func (r *ReduceSuite) TestSimpleNonKeyedReducePipeline() {
+	w := r.Given().Pipeline("@testdata/simple-non-keyed-reduce-pipeline.yaml").
+		When().
+		CreatePipelineAndWait()
+
+	defer w.DeletePipelineAndWait()
+
+	// wait for all the pods to come up
+	w.Expect().
+		VertexPodsRunning()
+
+	// port forward source vertex(to publish messages)
+	defer w.VertexPodPortForward("in", 8443, dfv1.VertexHTTPSPort).
+		TerminateAllPodPortForwards()
+
+	// publish messages to source vertex, with event time starting from 60000
+	startTime := 60000
+	for i := 0; i < 300; i++ {
+		eventTime := strconv.Itoa(startTime + i*1000)
+
+		HTTPExpect(r.T(), "https://localhost:8443").POST("/vertices/in").WithBytes([]byte("1")).WithHeader("X-Numaflow-Event-Time", eventTime).
+			Expect().
+			Status(204)
+		HTTPExpect(r.T(), "https://localhost:8443").POST("/vertices/in").WithBytes([]byte("2")).WithHeader("X-Numaflow-Event-Time", eventTime).
+			Expect().
+			Status(204)
+		HTTPExpect(r.T(), "https://localhost:8443").POST("/vertices/in").WithBytes([]byte("3")).WithHeader("X-Numaflow-Event-Time", eventTime).
+			Expect().
+			Status(204)
+	}
+
+	// since there is no key, all the messages will be assigned to same window
+	// the sum should be 60(since the window is 10s)
+	w.Expect().VertexPodLogContains("sink", "Payload -  60  Key -  NON_KEYED_STREAM  Start -  60000  End -  70000")
+}
+
+// two reduce vertex(keyed and non keyed)
+func (r *ReduceSuite) TestComplexReducePipelineKeyedNonKeyed() {
+	w := r.Given().Pipeline("@testdata/complex-reduce-pipeline.yaml").
+		When().
+		CreatePipelineAndWait()
+
+	defer w.DeletePipelineAndWait()
+
+	// wait for all the pods to come up
+	w.Expect().
+		VertexPodsRunning()
+
+	// port forward source vertex(to publish messages)
+	defer w.VertexPodPortForward("in", 8443, dfv1.VertexHTTPSPort).
+		TerminateAllPodPortForwards()
+
+	// publish messages to source vertex, with event time starting from 60000
+	startTime := 60000
+	for i := 0; i < 300; i++ {
+		eventTime := strconv.Itoa(startTime + i*1000)
+
+		HTTPExpect(r.T(), "https://localhost:8443").POST("/vertices/in").WithBytes([]byte("1")).WithHeader("X-Numaflow-Event-Time", eventTime).
+			Expect().
+			Status(204)
+		HTTPExpect(r.T(), "https://localhost:8443").POST("/vertices/in").WithBytes([]byte("2")).WithHeader("X-Numaflow-Event-Time", eventTime).
+			Expect().
+			Status(204)
+	}
+
+	// since the key can be even or odd and the first window duration is 10s(which is keyed)
+	// and the second window duration is 60s(non-keyed)
+	// the sum should be 180(60 + 120)
+	w.Expect().VertexPodLogContains("sink", "Payload -  180  Key -  NON_KEYED_STREAM  Start -  120000  End -  180000")
+}
+
+func (r *ReduceSuite) TestSimpleReducePipelineFailOverUsingWAL() {
+	w := r.Given().Pipeline("@testdata/simple-reduce-pipeline-wal.yaml").
 		When().
```

Comment:
@xdevxy `TestSimpleReducePipelineFailOverUsingWAL` test keeps failing, I have increased the number of input messages, but I am not sure if that will fix the issue. Can you please check?

Reply:
looks like increasing the number of input messages has fixed it.

Reply:
yes, we need to figure out "the why"

Reply:
opened #408